packages/flutter/lib/src/widgets/widget_span.dart

```diff
@@ -0,0 +1,254 @@
+/// overall [InlineSpan] tree will not mention [WidgetSpan]s unless
+/// [semanticsLabel] is provided. Semantics for the widget itself will still
+/// function independently from the [InlineSpan] semantics label.
+///
+/// See also:
+///
+///  * [TextSpan], a node that represents text in a [InlineSpan] tree.
+///  * [Text], a widget for showing uniformly-styled text.
+///  * [RichText], a widget for finer control of text rendering.
+///  * [TextPainter], a class for painting [InlineSpan] objects on a [Canvas].
+@immutable
+class WidgetSpan extends PlaceholderSpan {
+  /// Creates a [WidgetSpan] with the given values.
+  ///
+  /// The [child] property should be non-null. [WidgetSpan] is a leaf node in
+  /// the [InlineSpan] tree
+  ///
+  /// A [TextStyle] may be provided with the [style] property, but only the
+  /// decoration, foreground, background, and spacing options will be used.
+  const WidgetSpan({
+    @required this.child,
+    InlineWidgetAlignment alignment = InlineWidgetAlignment.bottom,
+    TextBaseline baseline,
+    TextStyle style,
+  }) : assert(child != null),
+       assert((alignment == InlineWidgetAlignment.aboveBaseline ||
+               alignment == InlineWidgetAlignment.belowBaseline ||
+               alignment == InlineWidgetAlignment.baseline) ? baseline != null : true),
+       super(alignment:
+           // Convert InlineWidgetAlignment to PlaceholderAlignment in a const fashion.
+           alignment == InlineWidgetAlignment.baseline ? ui.PlaceholderAlignment.baseline :
+           alignment == InlineWidgetAlignment.aboveBaseline ? ui.PlaceholderAlignment.aboveBaseline :
+           alignment == InlineWidgetAlignment.belowBaseline ? ui.PlaceholderAlignment.belowBaseline :
+           alignment == InlineWidgetAlignment.top ? ui.PlaceholderAlignment.top :
+           alignment == InlineWidgetAlignment.bottom ? ui.PlaceholderAlignment.bottom :
+           alignment == InlineWidgetAlignment.middle ? ui.PlaceholderAlignment.middle : null,
+         baseline: baseline, style: style,
+       );
+
+  /// The widget to embed inline within text.
+  final Widget child;
+
+  /// Adds a placeholder box to the paragraph builder if a size has been
+  /// calculated for the widget.
+  ///
+  /// Sizes are provided through [dimensions], which should contain a 1:1
+  /// in-order mapping of widget to laid out dimensions. If no such dimension
+  /// is provided, the widget will be skipped.
+  ///
+  /// Since widget sizes are calculated independently from the rest of the
+  /// paragraph, the [textScaleFactor] is ignored.
+  @override
+  void build(ui.ParagraphBuilder builder, { double textScaleFactor = 1.0, List<PlaceholderDimensions> dimensions }) {
+    assert(debugAssertIsValid());
+
+    final bool hasStyle = style != null;
+    if (hasStyle)
+      builder.pushStyle(style.getTextStyle(textScaleFactor: textScaleFactor));
+    if (dimensions != null) {
```

Comment:
shouldn't we assert `dimensions != null`?

Reply:
By using if, this allows a InlineSpan tree to be used for plain text without widgets and placeholders if no dimensions are provided. This can be useful by allowing reuse of the same InlineSpan tree when content does not need to include all of the complexities of widgets. By building without dimensions, the widget spans are skipped and produces something akin to a plaintext version of the InlineSpan tree.

If you think this is not behavior we would like to support, I can change it into an assert and document that dimensions is required if using WidgetSpans

Reply:
Do we have a customer asking for this?

Reply:
No, I will just enforce that dimensions must be provided. It'll prevent confusion over why the widgets arent rendering.